<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>Scope+ User Tutorial</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Scope+ User Tutorial";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> Scope+ User Tutorial
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href=".">Scope+ User Tutorial</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-file">Input file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#input-file-format">Input file format</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#database-collection">Database collection</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#dependencies">Dependencies</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#reimplementation-one-by-one-steps">Reimplementation one-by-one steps</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#quickstart-guide-after-data-imported-to-mongodb">Quickstart guide (after data imported to MongoDB)</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#prepare-your-data-for-scope-reimplementation">Prepare your data for Scope+ reimplementation</a>
    </li>
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">Scope+ User Tutorial</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Scope+ User Tutorial</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="scope-user-tutorial">Scope+ User Tutorial</h1>
<p>This tutorial is made for Scope+ and is straightforward. You only need three separate files to start implement your own single-cell atlas portal using the Scope+ architecture.</p>
<h2 id="input-file">Input file</h2>
<ul>
<li>Metadata</li>
<li>Count matrix</li>
<li>UMAP coordinates</li>
</ul>
<p>Example input files are provided for your reference. Input files needed to be in .csv format.</p>
<p>We assume that users have analyzed their data in one of the most popular software Seurat in R. You can output your three input files from your Seurat object.</p>
<ul>
<li>Metadata: your own metadata file in .csv format i.e. meta.csv</li>
<li>Count matrix: </li>
</ul>
<pre><code class="language-sh"># convert the sparse count matrix into dense matrix
# https://stackoverflow.com/questions/4558277/write-a-sparse-matrix-to-a-csv-in-r 
sparse2triples &lt;- function(m) {
  SM = summary(m)
  D1 = m@Dimnames[[1]][SM[,1]]
  D2 = m@Dimnames[[2]][SM[,2]]
  data.frame(row=D1, col=D2, x=m@x)
}
# get matrix file
dense_matrix &lt;- sparse2triples(srt_obj[[&quot;RNA&quot;]]@counts)

colnames(dense_matrix) &lt;- c(&quot;gene_name&quot;, &quot;barcode&quot;,&quot;expression&quot;)
write.csv(dense_matrix, file=&quot;matrix.csv&quot;, row.names=FALSE) 
</code></pre>
<p>For large-scale single-cell data we also provide utilities to convert the 10X single-cell RNA-seq data to MongoDB import collection .csv format at <a href="https://github.com/hiyin/covid19_cell_atlas_portal_supp/tree/main/utilities">here</a>.</p>
<ul>
<li>UMAP:
If you have run the UMAP step, otherwise please refer to Seurat UMAP method and run it first. </li>
</ul>
<pre><code class="language-sh">umap_coord &lt;- dplyr::as_tibble(data.frame(seurat_object@reductions$umap@cell.embeddings), rownames = &quot;id&quot;)
colnames(umap_coord) &lt;- c(”id“, “UMAP1”,”UMAP2”)
write.csv(umap_coord, file=”umap.csv”, row.names=FALSE)
</code></pre>
<h2 id="input-file-format">Input file format</h2>
<table>
 <td>**File name**

   </td>
   <td>**Collection name**

   </td>
   <td>**Columns**

   </td>
  <tr>
   <td>metadata.csv

   </td>
   <td>single_cell_meta_v4

   </td>
   <td>["id", "meta_age_category", "meta_sample_id2","meta_patient_id", "meta_dataset", "level2", "meta_severity", "meta_days_from_onset_of_symptoms", "meta_outcome", "meta_gender", "Country"]

   </td>
  </tr>
  <tr>
   <td>matrix.csv

   </td>
   <td>matrix

   </td>
   <td>["barcode", "gene_name","expression"]

   </td>
  </tr>
  <tr>
   <td>umap.csv

   </td>
   <td>umap

   </td>
   <td>["UMAP1","UMAP2", "id"]

   </td>
  </tr>
</table>

<h2 id="database-collection">Database collection</h2>
<table>
  <tr>
   <td>**Column name**

   </td>
   <td>**Value format**

   </td>
   <td>**Description**

   </td>
   <td>**Example**

   </td>
  </tr>
  <tr>
   <td colspan="4" >Collection schema: **single_cell_meta_v4**
   </td>
  </tr> 
   <td>id

   </td>
   <td>String

   </td>
   <td>Cell barcode (unique)

   </td>
   <td>
   </td>
  <tr>
   <td>meta_patient_id

   </td>
   <td>String

   </td>
   <td>Patient identifier

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_sample_id2

   </td>
   <td>String

   </td>
   <td>Sample identifier (one patient may have multiple samples)

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Country

   </td>
   <td>String

   </td>
   <td>Country origin of the dataset

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_age_category

   </td>
   <td>String

   </td>
   <td>Age category in intervals 

   </td>
   <td>e.g. “18-30”

   </td>
  </tr>
  <tr>
   <td>level2

   </td>
   <td>String

   </td>
   <td>Cell type predictions

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_severity

   </td>
   <td>String

   </td>
   <td>The severity of the symptoms regarding patient

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_dataset

   </td>
   <td>String

   </td>
   <td>Dataset origin

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_days_from_onset_of_symptoms

   </td>
   <td>Number

   </td>
   <td>Days from the onset of symptoms

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>meta_gender

   </td>
   <td>String

   </td>
   <td>Gender 

   </td>
   <td>i.e. “female” or “male”

   </td>
  </tr>
  <tr>
   <td>meta_outcome

   </td>
   <td>String

   </td>
   <td>Health outcome

   </td>
   <td> i.e. “diseased”, “discharged” etc.

   </td>
  </tr>
   <td colspan="3" >Collection schema: **umap**

   </td>
   <td>
   </td>
  <tr>
   <td>id

   </td>
   <td>String

   </td>
   <td>Cell barcode (unique)

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>UMAP1

   </td>
   <td>String

   </td>
   <td>X-coordinate of the cell

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>UMAP2

   </td>
   <td>String

   </td>
   <td>Y-coordiante of the cell

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td colspan="4" >Collection schema: **matrix**

   </td>
   </tr>
   <td>barcode

   </td>
   <td>String

   </td>
   <td>Cell barcode (unique)

   </td>
   <td>
   </td>
  <tr>
   <td>gene_name

   </td>
   <td>String

   </td>
   <td>Name of the gene expressed in the cell

   </td>
   <td>e.g. CD19

   </td>
  </tr>
  <tr>
   <td>expression

   </td>
   <td>Number

   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>

</table>

<h2 id="dependencies">Dependencies</h2>
<ul>
<li>MongoDB installed</li>
<li>Python3.9 installed</li>
<li>MongoDB Compass installed (optional, for visual operation of the database)</li>
<li>sqlite installed</li>
</ul>
<h2 id="reimplementation-one-by-one-steps">Reimplementation one-by-one steps</h2>
<ol>
<li>Install the system dependencies listed above, MongoDB, Python, MongoDB Compass (optional) and sqlite.</li>
<li>
<p>Create a database in MongoDB named cov19atlas_new, and create three collections namely under the database</p>
<ol>
<li>single_cell_meta_v4  </li>
<li>umap                </li>
<li>matrix              </li>
</ol>
</li>
<li>
<p>Import the three datasets into MongoDB using MongoDB Compass, make sure all data field during import needs to be in STRING format.: </p>
<ol>
<li>single_cell_meta_v4 (<a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_metadata.csv">importdata_meta.csv</a>)</li>
<li>umap (<a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_umap.csv">importdata_umap.csv</a>) and </li>
<li>matrix (<a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_matrix.csv">importdata_matrix.csv</a>).</li>
</ol>
</li>
<li>
<p>Download Covidscope web portal resources below into flask_resources directory under your $HOME, without them the web portal can't be initialized</p>
<ol>
<li><a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/features.tsv">features.tsv</a></li>
<li><a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/db.sqlite">db.sqlite</a></li>
</ol>
</li>
<li>
<p>Clone the repository, install the packages for Covidscope and run the web portal code</p>
</li>
</ol>
<h2 id="quickstart-guide-after-data-imported-to-mongodb">Quickstart guide (after data imported to MongoDB)</h2>
<pre><code class="language-sh">$ git clone https://github.com/hiyin/covid19_cell_atlas_portal.git
$ cd covid19_cell_atlas_portal

# create virtual environment
$ python3.9 -m pip install --upgrade pip
$ python3.9 -m venv venv # this installs the venv folder in the current directory
$ source venv/bin/activate
$ pip install -r requirements.txt

# deactivate the environment and reactivate it for a fresh start
$ deactivate
$ source venv/bin/activate

# initialize database download it and put it in flask_resources/ directory in yor $HOME
mv db.sqlite $HOME/flask_resources/
mv features.tsv $HOME/flask_resources/

# set environment
$ export FLASK_ENV=development
$ export FLASK_APP=manage.py

# launch
$ flask run
</code></pre>
<p>You will have local version of Scope+ running at 127.0.0.1:5000 by default.</p>
<p>You shall be able to have your own version of Scope+ running with your custom data files if you could modify to the data format same as the example files we used here for demonstration!</p>
<h2 id="prepare-your-data-for-scope-reimplementation">Prepare your data for Scope+ reimplementation</h2>
<p>We assume that you start with the two common files after you have collected your single-cell RNA-seq data i.e. meta data, and a count matrix folder in 10X single-cell sequencing format. </p>
<p>Below is a example pipeline to help you to process the files.
We asssume that you have a metadata file following our structures (if not please edit according to our example metadata)</p>
<ol>
<li>
<p>Example metadata:
    <a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_metadata.csv">metadata</a></p>
</li>
<li>
<p>Example 10X format matrix folder files:</p>
<ol>
<li><a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/raw/hoehn/barcodes.tsv.gz">barcodes.tsv.gz</a></li>
<li><a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/raw/hoehn/genes.tsv.gz">genes.tsv.gz</a></li>
<li><a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/raw/hoehn/matrix.mtx.gz">matrix.mtx.gz</a></li>
</ol>
</li>
</ol>
<p>Download them and save as into a new directory named hoehn_2021/</p>
<pre><code class="language-sh">meta &lt;- read.csv(&quot;importdata_metadata.csv&quot;)
# Prepare matrix db collection source
hoehn &lt;- Read10X(&quot;hoehn_2021/&quot;, gene.column = 1) # default tsv.gz files (downloaded from Covidscope)
srt_obj &lt;- CreateSeuratObject(hoehn)
rownames(meta) &lt;- meta$id
metadata_frame &lt;- srt_obj@meta.data
univ_meta &lt;- cbind(metadata_frame, meta)
srt_obj &lt;- AddMetaData(srt_obj, univ_meta)

# if the user didn't have pca
srt_obj &lt;- NormalizeData(srt_obj)
srt_obj &lt;- FindVariableFeatures(srt_obj, selection.method = &quot;vst&quot;, nfeatures = 2000)
all.genes &lt;- rownames(srt_obj)
srt_obj &lt;- ScaleData(srt_obj, features=all.genes)
srt_obj &lt;- RunPCA(srt_obj, features = VariableFeatures(object = srt_obj))
srt_obj &lt;- RunUMAP(srt_obj, dims = 1:40)

# Prepare UMAP file for database import 
umap_coord &lt;- dplyr::as_tibble(data.frame(srt_obj@reductions$umap@cell.embeddings
), rownames = &quot;id&quot;)
colnames(umap_coord) &lt;- c(&quot;id&quot;, &quot;UMAP1&quot;,&quot;UMAP2&quot;)
write.csv(umap_coord, file=&quot;importdata_umap.csv&quot;, row.names = FALSE)

# Prepare matrix file for database import 
# input: a sparse matrix with named rows and columns (dimnames)
# returns: a data frame representing triplets (r, c, x) suitable for writing to a CSV file
sparse2triples &lt;- function(m) {
  SM = summary(m)
  D1 = m@Dimnames[[1]][SM[,1]]
  D2 = m@Dimnames[[2]][SM[,2]]
  data.frame(row=D1, col=D2, x=m@x)
}

# get matrix file
towrite_matrix &lt;- sparse2triples(srt_obj[[&quot;RNA&quot;]]@counts)
colnames(towrite_matrix) &lt;- c(&quot;gene_name&quot;, &quot;barcode&quot;, &quot;expression&quot;)
# write matrix data
write.csv(towrite_matrix, file=&quot;importdata_matrix.csv&quot;, row.names=FALSE) 
</code></pre>
<p>For a quick reproduction of Covidscope local version, you could download our prepared and processed files and directly import them into the MongoDB:
1. <a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_matrix.csv">matrix</a>
2. <a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_metadata.csv">metadata</a>
3. <a href="https://covidscope-public-repository.s3.ap-east-1.amazonaws.com/user-tutorial/importdata_umap.csv">umap</a></p>
<h1 id="how-to-use-covidscope-data">How to use Covidscope data</h1>
<p>For any user would like to use Covidscope data, downloaded from https://covidsc.d24h.hk/data, Covidscope provide the data in 10X format therefore you can read the three .gz files using Seurat package's function <code>Read10X</code>. An example is given below</p>
<pre><code class="language-sh">data &lt;- Read10X(&quot;lee_2020/&quot;, gene.column = 1) # default tsv.gz files (downloaded from Covidscope)
srt_obj &lt;- CreateSeuratObject(data)
</code></pre>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>

<!--
MkDocs version : 1.5.3
Build Date UTC : 2023-10-25 15:05:54.577314+00:00
-->
